group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1830,6 +1832,63 @@ public void onLoaded() {
         });
     }
 
+    /**
+     * Called when the partition is unloaded. Cancel all existing timers for the group.
+     * ClassicGroup: Complete all awaiting join futures and sync futures. Transition group to Dead.
+     */
+    public void onUnloaded() {
+        groups.values().forEach(group -> {
+            switch (group.type()) {
+                case CONSUMER:
+                    ConsumerGroup consumerGroup = (ConsumerGroup) group;
+                    log.info("[GroupId={}] Unloading group metadata for group epoch {}.",
+                        consumerGroup.groupId(), consumerGroup.groupEpoch());
+
+                    consumerGroup.members().values().forEach(member -> {
+                        timer.cancel(consumerGroupSessionTimeoutKey(consumerGroup.groupId(), member.memberId()));
+                        timer.cancel(consumerGroupRevocationTimeoutKey(consumerGroup.groupId(), member.memberId()));
+                    });
+
+                    break;
+                case CLASSIC:
+                    ClassicGroup classicGroup = (ClassicGroup) group;
+                    log.info("[GroupId={}] Unloading group metadata for generation {}.",
+                        classicGroup.groupId(), classicGroup.generationId());
+
+                    classicGroup.transitionTo(DEAD);
+                    switch (classicGroup.previousState()) {
+                        case EMPTY:
+                        case DEAD:
+                            break;
+                        case PREPARING_REBALANCE:
+                            classicGroup.allMembers().forEach(member -> {
+                                classicGroup.completeJoinFuture(member, new JoinGroupResponseData()
+                                    .setMemberId(member.memberId())
+                                    .setMembers(Collections.emptyList())
+                                    .setGenerationId(NO_GENERATION)
```

Comment:
Members are empty by default and the generation id is -1 by default too. I wonder if we could just remove those two.

Reply:
will remove them. sorry for the confusion, we don't need https://issues.apache.org/jira/browse/KAFKA-16299. i don't know where i saw the default values but i remember they were set as 0 and null..

removed all NO_GENERATION and NO_PROTOCOL_NAME usages as they are the defaults